### PR TITLE
0719 create a json data adaptor

### PR DESCRIPTION
**Pull request by tforster**
The primary focus of this PR is the implementation of static JSON files instead of GraphQL queries. Static JSON files are expected to be found in the src/data directory (fka src/db). If the data specified in webproducer.yml is GraphQL then an option can be passed to WebProducer, currently, through the implementing project's index.js, that will tell WebProducer to save the results of the query to src/data directory. This is useful for grabbing a snapshot of data before working offline. The data definition in webproducer.yml will need to be changed from GraphQL to the local filesystem in order to use the snapshot. 

To make changing definitions in webproducer.yml easier, a new format has been introduced (breaking change!) that now supports aliases. Multiple data, templates and destination aliases can be set up in webproducer.yml and targeted via environment variables.

The above changes impacted a lot of files in the project. And along the way, the opportunity was taken for some long-needed renaming of certain modules. In most, if not all cases, references to Vinyl have been removed from module names and variables. In addition, module names are starting to align with the emerging concept of adapters. File adapters, data adapters, etc.

These changes were developed in parallel with a ~~first~~ second release of Wazzio's WazzCards www project as a suitable test case was required. It will be helpful to review PRs together.

@webdivelement Let's set up a time for a join review where I can take you through all the highlights.

### CODE REVIEW

File: src/Utils.js [hunk 1]
```diff
@@ -27,18 +34,95 @@ class Utils {
...
+   *
+   * Identifies the type of path as filesystem, S3 or HTTP 
+   * 
+   * @static
+   * @param {string} aPath: A path in the format s3://..., http(s)://... or *nix
+   * @returns {string}:     s3, http or file indicating the type
+   * @memberof Utils
+   */
+  static identifyFileSystem(aPath) {
+    //const pattern = /^(s3:)|^(http:|https:)|^(\/|\.\.\/|\.\/|\w*)/;
```
Comment (tforster) [inline src/Utils.js:83]:
```suggestion
```

Remove commented code

File: README.md [hunk 2]
```diff
@@ -36,53 +36,77 @@ Clone this repository as your new project `git clone git@github.com:tforster/web
...
-     acl: public-read # default
-     deleteOnSync: true # default
-     webserver: # optional
-       setContentTypeHeader: true # default
-       cloudFrontDistributionId: EZP3VGJ7GV1RO # optional
-
-    # preview is currently tightly bound to the DatoCMS implementation for specifying draft vs published
-    preview: ${env:PREVIEW}
+   # path: A pointer to the root of the data. E.g. a single file containing the data or a database via  connection string
+   # meta: A filesystem or S3 bucket path containing one or both of a query and transform module. Types graphql, sql and mongo all
```
Comment (webdivelement) [inline README.md:101]: Looking at some old files, I see I used `base` (and to a lesser extent `root`) for the name of the `path` property as noted above.  Then for the `meta` property above, the name `path` was used.  Not sure if this helps us get closer to the terminology we want.

File: src/GraphQLDataAdapter.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+class GraphQLDataAdapter {
+  /**
+   * Fetch wrapper around the provided GraphQL endpoint and GraphQL query
+   * @static
+   * @param {hash} options:     Includes GraphQL endpoint, DatoCMS token, etc
+   * @param {boolean} preview:  True to retrieve unpublished (aka draft) content
+   * @returns
+   * @memberof GraphQLDataSource
+   */
+  static async data(endpoint, query, token, published = false) {
```
Comment (webdivelement) [inline src/GraphQLDataAdapter.js:20]: The comment header could use an update based on the current parameters.

File: src/GraphQLDataAdapter.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    }
+  }
+
+  /**
+   *
+   * Rewrites the URL to account for provider specific syntax when selecting draft vs published items
+   * @static
+   * @param {*} provider
+   * @param {*} url
+   * @param {*} published
```
Comment (webdivelement) [inline src/GraphQLDataAdapter.js:62]: It appears that the parameters here do not have dynamic types, so the comment header could be updated to note `string`, `string` and `boolean`.

File: src/S3FileAdapter.js [hunk 5]
```diff
@@ -9,13 +9,12 @@ const micromatch = require("micromatch");
...
 
 // Project dependencies
-const Utils = require("./Utils");
 
 /**
  * Implements Gulp-like dest() function to terminate a piped process by uploading the supplied Vinyl file to AWS S3
  * @class VinylS3
  */
-class VinylS3 {
+class s3FileAdapter {
```
Comment (webdivelement) [inline src/S3FileAdapter.js:17]: Should this class name start with a capital?

File: src/Utils.js [hunk 6]
```diff
@@ -27,18 +34,94 @@ class Utils {
     return Promise.all(directories.map((directory) => fs.rmdir(directory, { recursive: true })));
   }
 
+
+  /**
+   *
+   *
+   * @static
+   * @param {} data
```
Comment (webdivelement) [inline src/Utils.js:42]: This comment header could be updated to include the description and the `data` parameter type.

File: src/Utils.js [hunk 7]
```diff
@@ -27,18 +34,94 @@ class Utils {
...
   /**
    * Returns the mimetype matching the supplied extension. Empty extensions are presumed to come from .html slugs.
    * @static
    * @param {string} ext: The extension, if available, to lookup the mime type for
    * @returns {string}:   The corresponding mimetype, or text/html if no extension was provided.
    * @memberof Utils
    */
-  static getMimeType(path) {
-    const mimeType = mime.getType(path);
+  static getMimeType(aPath) {
```
Comment (webdivelement) [inline src/Utils.js:67]: I see that this was changed from `path` to `aPath`, is there a specific reason for the naming of the parameter?  The comment header could also use an update to reflect the parameter name.

File: src/Utils.js [hunk 8]
```diff
@@ -27,18 +34,94 @@ class Utils {
...
+    const mimeType = mime.getType(aPath);
     return mimeType || "text/html";
   }
 
+
+  /**
+   *
+   * Identifies the type of path as filesystem, S3 or HTTP
```
Comment (webdivelement) [inline src/Utils.js:75]: Super nit-picky, however, there is an empty line at the beginning of the comment header.

File: src/index.js [hunk 9]
```diff
@@ -73,13 +93,28 @@ class WebProducer {
     return { srcStreamReadable, destStreamWritable, destStreamReadable };
   }
 
+  /**
+ 
+   * @memberof WebProducer
```
Comment (webdivelement) [inline src/index.js:98]: Is this an orphaned comment header?

File: src/Utils.js (hunk 7 above)
Comment (tforster) [reply src/Utils.js:67]: Because I added `const path = require("path");` after the fact. In hindsight `aPath` sucks so I just renamed to `pathname`.

**Pull request merged**